internal/pkgecosystem/npm.go

```diff
@@ -29,11 +32,127 @@ func getNPMLatest(pkg string) (string, error) {
 	return details.DistTags.Latest, nil
 }
 
+func makeNpmPackageSpec(pkgName, version string) string {
+	if version == "" {
+		return pkgName
+	} else {
+		return pkgName + "@" + version
+	}
+}
+
+// npmPackData represents the JSON format output by `npm pack`
+// when it successfully downloads a package archive.
+// The format was found by inspecting JSON output of the NPM CLI.
+type npmPackData struct {
+	PackageId    string        `json:"id"`
+	Name         string        `json:"name"`
+	Version      string        `json:"version"`
+	Size         int           `json:"size"`
+	UnpackedSize int           `json:"unpackedSize"`
+	ShaSum       string        `json:"shasum"`
+	Integrity    string        `json:"integrity"`
+	Filename     string        `json:"filename"`
+	Files        []npmFileData `json:"files"`
+	EntryCount   int           `json:"entryCount"`
+	Bundled      interface{}   `json:"bundled"` // not needed
+}
+
+// npmPackData represents the JSON format of a single file
+// listed in an NPM package archive, as part of npmPackData.
+// The format was found by inspecting JSON output of the NPM CLI.
+type npmFileData struct {
+	Path string `json:"path"`
+	Size int    `json:"size"`
+	Mode int    `json:"mode"`
+}
+
+// npmPackError represents the JSON format output by `npm pack`
+// when an error occurs while attempting to download a package archive.
+// The format was found by inspecting JSON output of the NPM CLI.
+type npmPackError struct {
+	Error struct {
+		Code    string `json:"code"`
+		Summary string `json:"summary"`
+		Detail  string `json:"detail"`
+	} `json:"error"`
+}
+
+func downloadNpmArchive(pkgName string, version string, directory string) (path string, err error) {
+	funcName := "downloadNpmArchive"
+	command := "npm"
+	if directory == "" {
+		// NPM defaults to current directory to download package archives, so we'll do the same
+		directory = "."
```

Comment:
We can be stricter here, and return an error instead, unless this behaviour is actually needed by callers?

Reply:
The comment makes it look like I'm being overly accommodating for underspecified inputs, but I actually just thought it would be useful to allow using the current directory for downloads. Do you think this shouldn't be allowed?

Reply:
Better to be explicit in these cases rather than introduce magic :) Callers can easily pass their own cwd if that's the behaviour they want.

Reply:
Okay yep, that's sound logic

Reply:
I'll return an error